### PR TITLE
Allow TrackerMuons with only an ME0 segment match

### PR DESCRIPTION
**Pull request by watson-ij**
#### PR description:

When only GEM segments from ME0 are matched to a track (e.g. at high eta where CSC doesn't cover), a TrackerMuon is not created because the MuonIdProducer only looks for DT or CSC segments when checking for a good TrackerMuon (requires at least one segment matched to the inner detector track). This PR adds GEM segments to the count of segments in MuonIdProducer, so that inner detector tracks which match to only an ME0 segment will pass the isGoodTrackerMuon and be saved into the reco muon collection.

#### PR validation:

Ran the phase-2 workflow 24811.0_TenMuExtendedE_0_200+2026D98 and checked before and after the changes that muons at high eta (|eta| > 2.4) which had a me0 segment match were not being saved before the changes, and are now being saved to the reco muon collection after the changes.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
@watson-ij   do we understand why the number of muons being registered in the DQM monitoring seem to be reduced by these changes?
See blue histograms w.r.t. black in wfs:  	23234.0, 24834.0 and 25034.999  on
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-10-24-2300+a0f421/59406/dqm-histo-comparison-summary.html

Thanks

**Comment by watson-ij** [comment]
@jfernan2 I think it should only add additional high eta muons, but I will check and see if I can figure out what happens in those workflows.

**Comment by jfernan2** [comment]
BTW: @watson-ij  this PR somehow relates to this long standing issue:
https://github.com/cms-sw/cmssw/issues/31424
I am not sure if we can take the occasion to move forward in that front. Thanks

**Comment by watson-ij** [comment]
I can only see the muons being added at high eta when doing some tests with 24834.0, but to be sure, I also stopped GE1/2 segments from being created, in case this had some affect. For the ME0 issue, yes, we should be able to start stripping out that code now. Once these issues of GE0 are worked out, I will start on that.

**Comment by watson-ij** [comment]
please test

**Comment by watson-ij** [comment]
I only checked rerunning step3. I will update so the HLT also runs (not sure about the addOnTests error though)

**Comment by watson-ij** [comment]
please test

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

**Comment by jfernan2** [comment]
@watson-ij   this comment seems to hold still
https://github.com/cms-sw/cmssw/pull/43107#issuecomment-1779332443

Do we finally understand why? Thanks

**Comment by watson-ij** [comment]
It looks like its the other way around, we get many many more muons with the pr. The GEMSegmentAndTrackArbitration doesn't actually check the arbitration flags, so accepts all trackermuons with a GE0 segment. Even though theres only a few segments, we get many matching tracks when pileup is on. I'm testing a change to the arbitration now to have the GEMSegmentAndTrackArbitration ensure that its the best muon candidate for the segment.

**Comment by watson-ij** [comment]
@jfernan2  I have a commit for this, but its a bit complicated, so I wanted to check what was happening with the TDR studies (did they also see hundreds of ME0Muons per event or not in PU studies, maybe I missed something), so wanted to run an older cmssw with ME0 (not GE0, as is now used). The last geometry version with an ME0 is D49, and seems to be last in CMSSW_12_5_5, but if I try to run the equivalent ttbar+pileup sample, theres no where that still keeps the old pileup events.

Is there a workflow that can produce the minbias for pileup locally instead of taking it from an existing sample? I seem to remember such a thing being possible but can't find it.

**Comment by jfernan2** [comment]
I am not sure which exact step you are looking for but, have you tried to seek using something like:  runTheMatrix.py -n --ext | grep TT | grep -v from

**Comment by watson-ij** [comment]
@jfernan2 sorry, I was mistaken, I reran a minbias sample and made it myself. It turns out that also in ME0 we were creating several dozen muons for each segment. I commited the changes that are able to arbitrate the segments, so only a single TrackerMuon (best in DeltaR) is produced for each ME0 segment. If there's a preference otherwise, or the solution is no good, please let me know.

**Comment by mandrenguyen** [comment]
please test

**Comment by jfernan2** [comment]
Hi @watson-ij 
It seems we are still losing some muons:
https://tinyurl.com/ymzovpgp
From the code point of view, the PR looks fine to me. 
From the physics point of view, I don't have the expertise to judge

**Comment by jfernan2** [comment]
No, blue is new, black is reference. All plots in those folders have less muon entries I am afraid

**Comment by jfernan2** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

**Review by missirol — COMMENTED** [review]
Unsolicited comments from a non-reviewer.

File: RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc [hunk 1]
```diff
@@ -27,6 +31,20 @@ GEMSegmentBuilder::GEMSegmentBuilder(const edm::ParameterSet& ps) : geom_(nullpt
 }
 GEMSegmentBuilder::~GEMSegmentBuilder() {}
 
+void GEMSegmentBuilder::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  edm::ParameterSetDescription ge0AlgoConfigDesc;
+  edm::ParameterSetDescription recAlgoConfigDesc;
+  desc.addOptional<bool>("enableGE0", true);
+  desc.addOptional<bool>("enableGE12", false);
```
[the remark this answers is not in the document]
Comment (watson-ij) [reply RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc:39]: Thanks for the comments. I made it this way because I was running into issues when running workflows that the HLT couldn't run because it didn't have this parameter, and I found the optional elsewhere. I will try the fillDescriptions suggestion, I guess it should resolve if theres a default value?

[the remark this answers is not in the document]
Comment (missirol) [reply RecoLocalMuon/GEMSegment/plugins/GEMSegmentBuilder.cc:39]: My understanding is that, if a plugin has a (static) `fillDescriptions` method with a parameter `add`ed with a default value, the plugin will use the default value of that parameter if said parameter is not explicitly included in the PSet of the python module (the given instance of the plugin in the configuration). Side note related to HLT: if the plugin has a `fillDescriptions` method, it's important that every parameter has a default value, because HLT normally uses the `cfi` produced via the `fillDescriptions` method to define which parameters are allowed for a given module (in the ConfDB GUI), and this works correctly on the HLT side only if a default value is specified (@cms-sw/hlt-l2 can correct me). Regarding `fillDescriptions` in general, here is a [link](https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp) with documentation.

In short, I would try something like https://github.com/missirol/cmssw/commit/9e5b8ea713b21aa8a5d20c63de7db7afe53e7686 (the existing `cfi` file is removed, because generated automatically via `GEMSegmentProducer::fillDescriptions`, hopefully with the same exact content).

**Comment by jfernan2** [comment]
Sure @mmusich 
We will do in the futuro this, we were note aware of this need until now

**Comment by watson-ij** [comment]
Thanks @missirol for the detailed explanation. One thing that was missing from what you sent is that the HLT python config also has a GEMDebug parameter for the GEMSegmentAlgorithm which doesn't seemed to be used at all. I added it into the fillDescriptions also to make a test workflow run.

**Comment by jfernan2** [comment]
+1

**Comment by mmusich** [comment]
unassign hlt
   * looks OK now

**Comment by srimanob** [comment]
+Upgrade

See changes in Phase-2 workflows, 23234.0, 24834.0, and 25034.999. This is expected from PR description.

**Comment by rappoccio** [comment]
+1
